File: 02_Testat/Java_Testat02/Testat02/Aufgabe2/Java/BankAccount.java
package Testat02.Aufgabe2.Java;

import java.util.concurrent.atomic.AtomicInteger;

// quickWithdraw mit atomaren Operationen ca. 3 mal schneller als withdraw mit do-while compareAndSet
public class BankAccount {
	private AtomicInteger balance = new AtomicInteger(0);

	public void deposit(int amount) {
		balance.getAndAdd(amount);
	}

	public boolean quickWithdraw(int amount) {
		if (amount <= balance.get()) {
			balance.getAndAdd(-amount);
			return true;
		} else {
			return false;
		}
	}

	public boolean withdraw(int amount) {
		int newValue = 0;
		int oldValue = 0;
		do {
			oldValue = balance.get();
			if (amount <= oldValue) {
				newValue = oldValue - amount;
			} else {
				return false;
			}
		} while (!balance.compareAndSet(oldValue, newValue));
		return true;
	}

	public int getBalance() {
		return balance.get();
	}
}

File: 02_Testat/DotNet_Testat02/02_DotNet/BankAccount/BankAccount.cs
using System.Runtime.InteropServices.WindowsRuntime;
using System.Threading;
using System.Threading.Tasks;

namespace BankAccount
{
    public class BankAccount
    {
        private volatile int _balance = 0;

        public void Deposit(int amount)
        {
            Interlocked.Add(ref _balance, amount);
        }

        public bool Withdraw(int amount)
        {
            if (amount <= _balance)
            {
                Interlocked.Add(ref _balance, -amount);
                return true;
            }

            return false;
        }

        public int GetBalance()
        {
            return _balance;
        }
    }
}